core/shared/src/main/scala/fs2/concurrent/Topic.scala

```diff
@@ -85,71 +88,136 @@ abstract class Topic[F[_], A] { self =>
+                  // _After_ we remove the bounded channel for this
+                  // subscriber, we need to drain it to unblock to
+                  // publish loop which might have already enqueued
+                  // something.
+                  def drainChannel: F[Unit] =
+                    subs.get(id).traverse_ { chan =>
+                      chan.close >> chan.stream.compile.drain
+                    }
+
+                  (subs - id, nextId) -> drainChannel
+                }.flatten >> subscriberCount.update(_ - 1)
+
+              Resource
+                .make(subscribe)(unsubscribe)
+                .as(chan.stream)
             }
 
-          def subscribeAwait(maxQueued: Int): Resource[F, Stream[F, A]] =
-            Resource
-              .eval(Q.bounded[F, A](maxQueued))
-              .flatMap { q =>
-                val subscribe = state.modify { case (subs, id) =>
-                  (subs.updated(id, q), id + 1) -> id
-                } <* subscriberCount.update(_ + 1)
-
-                def unsubscribe(id: Long) =
-                  state.modify { case (subs, nextId) =>
-                    // _After_ we remove the bounded queue for this
-                    // subscriber, we need to drain it to unblock to
-                    // publish loop which might have already enqueued
-                    // something.
-                    def drainQueue: F[Unit] =
-                      subs.get(id).traverse_ { q =>
-                        q.tryTake.iterateUntil(_.isEmpty)
-                      }
-
-                    (subs - id, nextId) -> drainQueue
-                  }.flatten >> subscriberCount.update(_ - 1)
-
-                Resource
-                  .make(subscribe)(unsubscribe)
-                  .as(Stream.fromQueueUnterminated(q))
-              }
+        def publish: Pipe[F, A, INothing] = { in =>
+          (in ++ Stream.exec(close.void))
+            .evalMap(publish1)
+            .takeWhile(_.isRight)
+            .drain
+        }
 
-          def publish: Pipe[F, A, Nothing] =
-            _.evalMap(publish1).drain
+        def subscribe(maxQueued: Int): Stream[F, A] =
+          Stream.resource(subscribeAwait(maxQueued)).flatten
 
-          def subscribe(maxQueued: Int): Stream[F, A] =
-            Stream.resource(subscribeAwait(maxQueued)).flatten
+        def subscribers: Stream[F, Int] = subscriberCount.discrete
 
-          def subscribers: Stream[F, Int] = subscriberCount.discrete
-        }
+        def close: F[Either[Topic.Closed, Unit]] =
+          signalClosure
+            .complete(())
+            .flatMap { completedNow =>
+              val result = if (completedNow) Topic.rightUnit else Topic.closed
+
+              state.get
+                .flatMap { case (subs, _) =>
+                  subs.foldLeft(F.unit) { case (op, (_, chan)) =>
+                    op >> chan.close.void
+                  }
+                }
+                .as(result)
+            }
```

Comment:
```suggestion
          (for {
            completedNow <- signalClosure.complete(())
            stt <- state.get
             _ <- stt._2.traverse_(_.close)
            } yield if (completedNow) Topic.rightUnit else Topic.closed
            ).uncancelable
```

Reply:
Oh, just realised there is no `traverse_` for `LongMap`. Perhaps we could add a function 

```Scala
  def allChannels(f: Channel[F, A] => F[_]): F[Unit] = 
    state.get.flatMap { case (subs, _) => 
      subs.foldLeft(F.unit){ case (op, (_, chan)) => op >> f(chan) }
    } 
```